### PR TITLE
[JENKINS-49707] Allow `node` blocks from deleted pods to be retried

### PR DESCRIPTION
**Pull request by jglick**
Downstream of https://github.com/jenkinsci/workflow-api-plugin/pull/217, https://github.com/jenkinsci/workflow-durable-task-step-plugin/pull/231, & https://github.com/jenkinsci/workflow-basic-steps-plugin/pull/203. Main functionality (without support for Declarative syntax) extracted from #1083, and without cross-restart tests.

### CODE REVIEW

**Review by jtnord — APPROVED** [review]

**Review by fcojfernandez — APPROVED** [review]

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/KubernetesAgentErrorCondition.java [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+        for (FlowNode callStack : scanner) {
+            WorkspaceAction ws = callStack.getPersistentAction(WorkspaceAction.class);
+            if (ws != null) {
+                String node = ws.getNode();
+                Node n = Jenkins.get().getNode(node);
+                if (n != null) {
+                    if (!(n instanceof KubernetesSlave)) {
+                        LOGGER.fine(() -> node + " was not a K8s agent");
+                        return false;
+                    }
```
Comment (jglick) [inline src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/KubernetesAgentErrorCondition.java:96]: https://github.com/jenkins-infra/pipeline-library/pull/405#discussion_r916058375

File: pom.xml [hunk 2]
```diff
@@ -104,6 +104,7 @@
     <dependency>
       <groupId>org.jenkins-ci.plugins.workflow</groupId>
       <artifactId>workflow-api</artifactId>
+      <version>1182.v41475e53ea_43</version> <!-- TODO until in BOM -->
```
Comment (jglick) [inline pom.xml:107]: https://github.com/jenkinsci/bom/pull/1252

**Review by Vlatombe — APPROVED** [review]

**Comment by jglick** [comment]
Still a couple tweaks I want to make here.

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/KubernetesAgentErrorCondition.java [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+    public boolean test(Throwable t, StepContext context) throws IOException, InterruptedException {
+        if (context == null) {
+            LOGGER.fine("Cannot check error without context");
+            return false;
+        }
+        if (!new AgentErrorCondition().test(t, context)) {
+            if (t instanceof FlowInterruptedException && ((FlowInterruptedException) t).getCauses().stream().anyMatch(ExecutorStepExecution.QueueTaskCancelled.class::isInstance)) {
+                LOGGER.fine(() -> "QueueTaskCancelled normally ignored by AgentErrorCondition but might be delivered here from Reaper.TerminateAgentOnContainerTerminated");
+                // TODO cleaner to somehow suppress that QueueTaskCancelled and let the underlying RemovedNodeCause be delivered
+                // (or just let AgentErrorCondition trigger on QueueTaskCancelled)
```
Comment (jglick) [inline src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/KubernetesAgentErrorCondition.java:72]: https://github.com/jenkinsci/workflow-durable-task-step-plugin/pull/249

**Review by dduportal — APPROVED** [review]

**Comment by jglick** [comment]
@Vlatombe ready for release from my perspective. ci.jenkins.io is already running an earlier version of this.

**Review by Vlatombe — APPROVED** [review]

**Pull request merged**